### ISSUE TITLE
can't make training with c++ API work

### ISSUE THREAD

**Issue by julienlopez** [Microsoft/CNTK#3083]
Hi, I'm trying to do some basic training with the C++ API but I can't seem to make it work.

I must be missing something obvious, but I can't find it, anyone has a idea what I'm doing wrong here?

```
using ElemType = float;

auto FullyConnectedLinearLayer(CNTK::Variable input, size_t outputDim, const CNTK::DeviceDescriptor& device,
	const std::wstring& outputName = L"", unsigned long seed = 1)
{
	assert(input.Shape().Rank() == 1);
	size_t inputDim = input.Shape()[0];

	auto timesParam = CNTK::Parameter(
		{ outputDim, inputDim }, CNTK::DataType::Float,
		CNTK::GlorotUniformInitializer(CNTK::DefaultParamInitScale, CNTK::SentinelValueForInferParamInitRank,
			CNTK::SentinelValueForInferParamInitRank, seed),
		device, L"timesParam");
	auto timesFunction = Times(timesParam, input, L"times");

	auto plusParam = CNTK::Parameter({ outputDim }, 0.0f, device, L"plusParam");
	return Plus(plusParam, timesFunction, outputName);
}

auto FullyConnectedDNNLayer(CNTK::Variable input, size_t outputDim, const CNTK::DeviceDescriptor& device,
	const std::function<CNTK::FunctionPtr(const CNTK::FunctionPtr&)>& nonLinearity,
	const std::wstring& outputName = L"", unsigned long seed = 1)
{
	return nonLinearity(FullyConnectedLinearLayer(input, outputDim, device, outputName, seed));
}

auto FullyConnectedFeedForwardClassifierNet(
	CNTK::Variable input, size_t numOutputClasses, size_t hiddenLayerDim, size_t numHiddenLayers,
	const CNTK::DeviceDescriptor& device,
	const std::function<CNTK::FunctionPtr(const CNTK::FunctionPtr&)>& nonLinearity, const std::wstring& outputName,
	unsigned long seed = 1)
{
	assert(numHiddenLayers >= 1);
	auto classifierRoot = FullyConnectedDNNLayer(input, hiddenLayerDim, device, nonLinearity, L"", seed);
	for (size_t i = 1; i < numHiddenLayers; ++i)
		classifierRoot = FullyConnectedDNNLayer(classifierRoot, hiddenLayerDim, device, nonLinearity, L"", seed);

	auto outputTimesParam = CNTK::Parameter({ numOutputClasses, hiddenLayerDim }, CNTK::DataType::Float,
		CNTK::UniformInitializer(0.5, seed), device);
	return Times(outputTimesParam, classifierRoot, 1, outputName);
}

void PrintTrainingProgress(const CNTK::TrainerPtr trainer, size_t minibatchIdx, size_t outputFrequencyInMinibatches)
{
	if ((minibatchIdx % outputFrequencyInMinibatches) == 0 && trainer->PreviousMinibatchSampleCount() != 0)
	{
		double trainLossValue = trainer->PreviousMinibatchLossAverage();
		double evaluationValue = trainer->PreviousMinibatchEvaluationAverage();
		printf("Minibatch %d: loss = %.8g, Evaluation criterion = %.8g\n", (int)minibatchIdx,
			trainLossValue, evaluationValue);
	}
}

inline bool getVariableByName(std::vector<CNTK::Variable> variableLists, std::wstring varName, CNTK::Variable& var)
{
	for (std::vector<CNTK::Variable>::iterator it = variableLists.begin(); it != variableLists.end(); ++it)
	{
		if (it->Name().compare(varName) == 0)
		{
			var = *it;
			return true;
		}
	}
	return false;
}

inline bool getOutputVariableByName(CNTK::FunctionPtr evalFunc, std::wstring varName, CNTK::Variable& var)
{
	return getVariableByName(evalFunc->Outputs(), varName, var);
}

std::vector<ElemType> createInputVector(const std::size_t nb_inputs)
{
	std::vector<ElemType> input_data(nb_inputs);
	std::iota(begin(input_data), end(input_data), 0.f);
	return input_data;
}

void eval(CNTK::FunctionPtr model, const std::size_t input_size, CNTK::Variable input_var, CNTK::Variable output_var, CNTK::DeviceDescriptor device)
{
	const std::size_t number_of_samples = 1; // batch of 1

	std::vector<ElemType> inputs = createInputVector(input_size);
	CNTK::NDShape inputShape = input_var.Shape().AppendShape({ 1, number_of_samples });
	CNTK::ValuePtr inputValue = CNTK::MakeSharedObject<CNTK::Value>(
		CNTK::MakeSharedObject<CNTK::NDArrayView>(inputShape, inputs, true));

	CNTK::ValuePtr outputValue;
	std::unordered_map<CNTK::Variable, CNTK::ValuePtr> outputs = { { output_var, outputValue } };
	model->Evaluate({ { input_var, inputValue } }, outputs, device);

	outputValue = outputs[output_var];
	CNTK::NDShape outputShape = output_var.Shape().AppendShape({ 1, number_of_samples });
	std::vector<ElemType> outputData(outputShape.TotalSize());
	CNTK::NDArrayViewPtr cpuArrayOutput
		= CNTK::MakeSharedObject<CNTK::NDArrayView>(outputShape, outputData, false);
	cpuArrayOutput->CopyFrom(*outputValue->Data());

	std::wcout << outputData[0] << " , " << outputData[1] << " , " << outputData[2] << std::endl;
}

int main()
{
	const std::wstring c_input_var_name = L"input";
	const std::wstring c_output_var_name = L"output";

	const std::size_t nb_inputs = 10;
	const std::size_t nb_outputs = 3;

	auto device = CNTK::DeviceDescriptor::UseDefaultDevice();
	auto input_var = CNTK::InputVariable({ nb_inputs }, CNTK::DataType::Float, c_input_var_name);
	auto model = FullyConnectedFeedForwardClassifierNet(input_var, nb_outputs, 512, 2, device, std::bind(CNTK::Sigmoid, std::placeholders::_1, L""), c_output_var_name);
	CNTK::Variable output_var;
	getOutputVariableByName(model, c_output_var_name, output_var);


	auto trainingLoss = SquaredError(model, output_var, L"lossFunction");
	auto prediction = SquaredError(model, output_var, L"predictionError");

	CNTK::LearningRateSchedule learningRatePerSample = 1;
	auto trainer = CNTK::CreateTrainer(model, trainingLoss, prediction, { CNTK::SGDLearner(model->Parameters(), learningRatePerSample) });

	for (std::size_t epoch = 0; epoch < 20; epoch++)
	{
		const std::vector<ElemType> inputData = createInputVector(nb_inputs);
		auto inputDataValue = CNTK::Value::CreateBatch(input_var.Shape(), inputData, device, true);

		const std::vector<ElemType> outputData{ 0., 1., 0. };
		auto outputDataValue = CNTK::Value::CreateBatch(output_var.Shape(), outputData, device, true);

		const auto res = trainer->TrainMinibatch({ { input_var, inputDataValue },{ output_var, outputDataValue } }, device);
		const size_t outputFrequencyInMinibatches = 1;
		PrintTrainingProgress(trainer, 0, outputFrequencyInMinibatches);

		std::wcout << "eval : " << std::endl;
		eval(model, nb_inputs, input_var, output_var, device);
	}
}
```

Loss is zero all the time, even after the first epoch, and the eval output is always the same

Moreover, I get that warning "WARNING: Function::Forward provided values for (1) extra arguments which are not required for evaluating the specified Function outputs!"

**Comment by KeDengMS** [comment]
This seems wrong:
```
	auto trainingLoss = SquaredError(model, output_var, L"lossFunction");
	auto prediction = SquaredError(model, output_var, L"predictionError");
```

The error should be measured on label. Here is measuring the squared error on the model output itself, which would always be zero.

**Comment by julienlopez** [comment]
yup, that was the problem, I created a new variable for the labels and it works much better, thanks
